frontend/lib/resolution/Resolver.cpp

```diff
@@ -2651,113 +2734,139 @@ void Resolver::resolveTupleUnpackDecl(const TupleDecl* lhsTuple,
+                        initExprT);
+  }
+
+  resolveTupleDecl(td, std::move(qt));
+}
+
+void Resolver::resolveTupleDecl(const TupleDecl* td, QualifiedType useType) {
   if (scopeResolveOnly) {
     return;
   }
 
   QualifiedType::Kind declKind = (Qualifier) td->intentOrKind();
-  QualifiedType useT;
+  QualifiedType qt;
 
   // Non-default intents are currently not allowed for tuple-grouped formals.
   //
   // Note: This doesn't apply to nested TupleDecls like ``(a, b)`` in the
   // formal ``((a, b), c)``. Such nested formals should be handled by
   // ``resolveTupleUnpackDecl``.
   if (td->isTupleDeclFormal() && declKind != QualifiedType::DEFAULT_INTENT) {
-    useT = QualifiedType(declKind, ErroneousType::get(context));
-    ResolvedExpression& result = byPostorder.byAst(td);
-    result.setType(useT);
+    qt = QualifiedType(declKind, ErroneousType::get(context));
+    auto& result = byPostorder.byAst(td);
+    result.setType(qt);
     return;
   }
 
-  // Figure out the type to use for this tuple
-  if (useType != nullptr) {
-    useT = QualifiedType(declKind, useType);
-  } else if (substitutions != nullptr &&
-             substitutions->count(td->id()) != 0) {
+  // If there is a substitution, use that instead.
+  if (substitutions != nullptr && substitutions->count(td->id()) != 0) {
     auto sub = substitutions->find(td->id())->second;
     auto subTup = sub.hasTypePtr() ? sub.type()->toTupleType() : nullptr;
 
     if (subTup == nullptr) {
       // Rely on hasTypePtr check below to recognize error
-      useT = QualifiedType();
+      qt = QualifiedType();
     } else {
-      useT = sub;
+      qt = sub;
     }
-  } else {
-    QualifiedType typeExprT;
-    QualifiedType initExprT;
 
-    auto typeExpr = td->typeExpression();
-    auto initExpr = td->initExpression();
-
-    if (typeExpr == nullptr && initExpr == nullptr) {
-      // Note: we seem to rely on tuple components being 'var', and relying on
-      // the tuple's kind instead. Without this, the current instantiation
-      // logic won't allow, for example, passing (1, 2, 3) to (?, ?, ?).
-      std::vector<QualifiedType> eltTypes;
-      for (auto decl : td->decls()) {
-        eltTypes.push_back(QualifiedType(QualifiedType::VAR,
-                                         getAnyType(*this, decl->id())));
-      }
-      auto tup = TupleType::getQualifiedTuple(context, eltTypes);
-      useT = QualifiedType(declKind, tup);
-    } else {
-      if (typeExpr != nullptr) {
-        ResolvedExpression& result = byPostorder.byAst(typeExpr);
-        typeExprT = result.type();
-      }
-      if (initExpr != nullptr) {
-        ResolvedExpression& result = byPostorder.byAst(initExpr);
-        initExprT = result.type();
-      }
-
-      useT = getTypeForDecl(td, typeExpr, initExpr,
-                            declKind, typeExprT, initExprT);
-
-    }
+  // Otherwise use the type that was provided.
+  } else {
+    qt = useType;
```

Comment:
notably, this inversion of logic (passing `useType` from outside instead of computing it in the `else` branch) means that we run the `useType`-building logic in the other overload even when this information is discarded when substitutions are found.